compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerToGlobalLoads.cpp

```diff
@@ -0,0 +1,381 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cstdint>
+#include <numeric>
+#include <optional>
+#include "iree/compiler/Codegen/Common/GPU/Passes.h"
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h"
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.h"
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUOps.h"
+#include "iree/compiler/Codegen/Utils/GPUUtils.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Arith/Utils/Utils.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/MemRef/Utils/MemRefUtils.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Utils/Utils.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Utils/StaticValueUtils.h"
+#include "mlir/Dialect/Vector/Transforms/VectorTransforms.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/OpDefinition.h"
+
+#define DEBUG_TYPE "iree-codegen-gpu-lower-to-global-loads"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_GPULOWERTOGLOBALLOADSPASS
+#include "iree/compiler/Codegen/Common/GPU/Passes.h.inc"
+
+namespace {
+//====---------------------------------------------------------------------===//
+// TODO: Move this to a common place.
+//====---------------------------------------------------------------------===//
+
+static constexpr int kNumBitsPerCopy = 32;
+
+// Slice the tensor into numParts parts, and return the i-th part.
+static std::optional<Value> sliceTensor(RewriterBase &rewriter, Value tensor,
+                                        size_t numParts, Value index) {
+  auto tensorType = cast<RankedTensorType>(tensor.getType());
+  auto outermostDimSize = tensorType.getDimSize(0);
+  if (outermostDimSize % numParts != 0) {
+    return std::nullopt;
+  }
+
+  // Create an extract slice
+  SmallVector<int64_t> newShape(tensorType.getShape());
+  newShape[0] = outermostDimSize / numParts;
+
+  auto loc = tensor.getLoc();
+  SmallVector<Value> newShapeOfSlice;
+  for (auto dim : newShape) {
+    newShapeOfSlice.push_back(
+        rewriter.create<arith::ConstantIndexOp>(loc, dim));
+  }
```

Comment:
This is unsued, and I think you may look for `tensor::getMixedSizes(rewriter, loc, tensor)`.